src/validators/StringValidator.ts

```diff
@@ -29,6 +37,28 @@ export class StringValidator<T extends string> extends BaseValidator<T> {
 		return this.addConstraint(stringLengthNe(length) as IConstraint<T>);
 	}
 
+	public email(): this {
+		const emailRegex = /^(([^<>()[\]\.,;:\s@\"]+(\.[^<>()[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i;
+		return this.addConstraint(stringRegex(emailRegex, 'email') as IConstraint<T>);
+	}
+
+	public url(allowedProtocols: string[] = ['https?']): this {
+		const urlRegex = new RegExp(`^(?:${allowedProtocols.join('|')}):\/\/[^\s\.]+\.\\S{2,}$`);
+		return this.addConstraint(stringRegex(urlRegex, 'url') as IConstraint<T>);
```

Comment:
This is vulnerable to accidental ReDoS injection due to unsanitised user-defined RegExp, and security tools won't pick up this kind of stuff.

Reply:
Also, it should validate using `new URL(input)`, not a RegExp, no matter how "slow" it is, it's the fastest option among full WHATWG-compliant URL parsers, and it's also built-in.

Reply:
changed it also added allowedDomains.